accumulator/pollardproof.go

```diff
@@ -208,3 +209,63 @@ func (p *Pollard) verifyBatchProof(
 
 	return true, proofmap
 }
+
+// CacheSim is a type capable of simulating pollard caching/remembering.
+type CacheSim struct {
+	// A bit set to mark cached positions.
+	// This is not optimal since it has a spatial complexity of O(n) where n is the number of UTXOs
+	// BUT only 1 bit per UTXO.
+	// currently(#643584) there are 66,391,686 UTXOs that means size(cached) = ~8MB.
+	cached *big.Int
+	// The number of leaves in the simulator.
+	numLeaves uint64
+}
+
+// NewCacheSimulator initialises a new cache simulator.
+func NewCacheSimulator(numLeaves uint64) *CacheSim {
+	return &CacheSim{cached: big.NewInt(0), numLeaves: numLeaves}
+}
+
+// Simulate simulates one modification to the pollard.
+// Takes a slice of target positions and a slice of bools representing new leaves (true means "cache the leaf").
+// Returns the proof positions that are needed to prove the targets (including the unkown targets).
+func (c *CacheSim) Simulate(targets []uint64, adds []bool) []uint64 {
+	// Figure out which targets are known/cached and which aren't.
+	var knownTargets, unknownTargets []uint64
+	for _, target := range targets {
+		if c.cached.Bit(int(target)) > 0 {
```

Comment:
Nit since it'll be eons until we get to 1 << 63 leaves but this could error out right?

https://play.golang.org/p/wo0-tE8s1KN

Reply:
@kcalvinalvin are you saying it will be the int casting that overflows? in that case would it be a problem on 32-bit platforms? are those supposed to be supported?

Reply:
on a 64-bit system it would cast to a signed int64 and `target` is unsigned, so the casted int could be negative

Reply:
i don't know what happens on a 32-bit system with all the uint64s we are using 😄